source/agora/consensus/SlashPolicy.d

```diff
@@ -169,6 +175,48 @@ public class SlashPolicy
         else
             return false;
     }
+
+    /***************************************************************************
+
+        Check if information for pre-images and slashed validators is valid
+
+        Params:
+            height = the height of proposed block
+            missing_validators = list of indices to the validator UTXO set
+                which have not revealed the preimage
+
+        Returns:
+            `null` if the information is valid at the proposed height,
+            otherwise a string explaining the reason it is invalid.
+
+    ***************************************************************************/
+
+    public string isInvalidPreimageRootReason (Height height,
```

Comment:
In the future it would be more efficient to construct the local array **once**, each time a new preimage is received. That way you can cache it and then the comparison is simple:

```
remote_missing == local_missing
```

Reply:
> In the future it would be more efficient to construct the local array once, each time a new preimage is received. That way you can cache it and then the comparison is simple:
> 
> remote_missing == local_missing

Sure, I will. That's one of the requirements described on the issue page.